lib/puppet/indirector/report/json.rb

```diff
@@ -0,0 +1,34 @@
+require 'puppet/transaction/report'
+require 'puppet/indirector/json'
+
+class Puppet::Transaction::Report::Json < Puppet::Indirector::JSON
+  include Puppet::Util::SymbolicFileMode
+
+  desc "Store last report as a flat file, serialized using JSON."
+
+  # Force report to be saved there
+  def path(name,ext='.json')
+    Puppet[:lastrunreport]
```

Comment:
Sorry, I realized there may be a problem with this after merging...

By default, puppetserver's report terminus is `:processor` and it's [`save` method](https://github.com/puppetlabs/puppet/blob/c2b87527385516b64fbe3ef991ea4417df346e9e/lib/puppet/indirector/report/processor.rb#L13) is called to handle the report. The processor walks the list of `Puppet[:reports]` processors giving each one a chance to process the report. The `:store` report processor is enabled by default, and it saves the report in a directory relative to [`Puppet[:reportdir]`](https://github.com/puppetlabs/puppet/blob/main/lib/puppet/reports/store.rb#L18). So server-side the yaml/json report termini are not used. It's possible to configure it via `Puppet[:routes_file]` but I don't think that's typical for reports.

By default, the agent's report terminus is `:rest` and its cache terminus is `:yaml`. Commit ccc944f21a259f0216b0bfd4873c98d89127a753 is where the agent started using the cache terminus to save a report to `lastrunreport`.

If we default the `json` (and `msgpack`) report termini to using `Puppet[:lastrunreport]`, then we end up storing the report in a file named `last_run_report.yaml`, even though it could contain `json` or `msgpack`. I'm wondering if we should instead change the name based on which terminus is being used?

I was thinking we might be able to do something like:

```ruby
    :lastrunreport =>  {
      :default  => lambda {
        if Puppet::Transaction::Report.indirection.cache_class == :json
          "$statedir/last_run_report.json"
        else
          "$statedir/last_run_report.yaml"
        end
      },   
```

However, the setting is evaluated during global settings initialization before the agent application configures the cache_class... I also don't like how the setting changes depending on which application is loaded, e.g. when running `puppet config print lastrunreport` won't be correct.

Reply:
Ah the agent hard codes the report cache terminus to yaml https://github.com/puppetlabs/puppet/blob/b99fbbeac63db4e20437ee34dcac47bd3ac08eff/lib/puppet/application/agent.rb#L476. It may be possible to override that using Puppet[:routes_file] with something like:

—-
agent:
   report:
      cache: rest
      terminus: json

If that is possible, maybe just update the Puppet[:lastrunreport] setting to mention if the report cache terminus is modified for agent apply or device applications then the file extension should be modified too, eg $statedir/last_run_report.json?